source/agora/common/BitMask.d

```diff
@@ -62,27 +62,32 @@ public struct BitMask
     }
 
     /// Bytes to hold bits to indicate if a validator has signed
-    private ubyte[] bytes;
+    private ubyte[] _bytes;
+
+    /// readonly public access to bytes
+    public const(ubyte[]) bytes() const scope @nogc
+    {
+        return this._bytes;
+    }
 
     public this (size_t length) inout
     {
         this.length = length;
         if (length > 0)
-            this.bytes = new inout(ubyte)[1 + ((length - 1) / 8)];
+            this._bytes = new inout(ubyte)[1 + ((length - 1) / 8)];
     }
 
     public this (size_t length, in ubyte[] bytes)
     {
         this(length);
-        this.bytes[] = bytes;
+        this._bytes[] = bytes;
     }
 
-    // set the bits that are set in given BitMask
-    public auto opOpAssign (string op : "|") (in BitMask add)
+    // copy bits set from given BitMask bytes
+    public void copyFrom (in BitMask rhs)
     {
-        assert(this.length == add.length, "BitMask assignment must be for same bit length");
-        this.bytes[] |= add.bytes[];
-        return this;
+        assert(this.length == rhs.length);
+        this._bytes[] = rhs.bytes;
     }
```

Comment:
So I did the change locally to see what would go wrong. And the issue is:
```D
    Block block;

    block.header.prev_block = prev_block.header.hashFull();
    block.header.height = prev_block.header.height + 1;
    block.header.time_offset = time_offset;
    block.header.random_seed = random_seed;
    block.header.validators = BitMask(validators);
    block.header.enrollments = enrollments;
```

Currently in `makeNewBlock`, in `agora.consensus.data.Block`. So the solution is either to make this kind of block building illegal, and force the user to do something like:
```D
Block block = Block(BlockHeader(...));
```
Or:
```D
Block block = { header: { validators: BitMask(validators) } };
```

Which I don't think is very good, because it feels like a pointless restriction.
Another approach would be to special-case the `length == 0` check to allow it (after all, there is no case where a default-initialized BitMask would be usable otherwise). The third approach is this. I guess this makes the most sense, so 👍 , but it would have been nice if it came with an explanation as to why the `opAssign` approach wasn't going to work.

Reply:
Whilst you were reviewing I started to create a very simple example to show the problem. Here is the gist https://gist.github.com/run-dlang/ec89c7b759785eb5770fdb359a3eabb9

Reply:
I will update the commit to include a better explanation.